src/libappimage/desktop_integration/integrator/Integrator.cpp

```diff
@@ -187,27 +188,114 @@ namespace appimage {
+                                Logger::warning("Using " + dirIconPath + " as default app icon");
+                                deployApplicationIcon(desktopEntryIconName, iconData);
+                            } catch (const PayloadIteratorError& error) {
+                                Logger::error(error.what());
+                                Logger::error("No icon was generated for: " + appImage.getPath());
+                            }
                         } catch (const PayloadIteratorError& error) {
                             Logger::error(error.what());
                             Logger::error("No icon was generated for: " + appImage.getPath());
                         }
-                    } else {
-                        // Generate the target paths were the Desktop Entry icons will be deployed
-                        std::map<std::string, std::string> iconFilesTargetPaths;
-                        for (const auto& itr: iconPaths)
-                            iconFilesTargetPaths[itr] = generateDeployPath(itr).string();
+                        return;
+                    }
+
+                    // Find the largest size icon
+                    std::string largestIconPath;
+                    int largestSize = 0;
+                    for (const auto& path : hicolorIconPaths) {
+                        // Extract size from path like "usr/share/icons/hicolor/512x512/apps/"
+                        std::string sizeStr;
+                        size_t sizePos = path.find("/hicolor/");
+                        if (sizePos != std::string::npos) {
+                            sizePos = path.find('/', sizePos + 9);
+                            if (sizePos != std::string::npos) {
+                                size_t endPos = path.find('/', sizePos + 1);
+                                if (endPos != std::string::npos) {
+                                    sizeStr = path.substr(sizePos + 1, endPos - sizePos - 1);
+                                    // Handle NxN format
+                                    size_t xPos = sizeStr.find('x');
+                                    if (xPos != std::string::npos) {
+                                        try {
+                                            int size = std::stoi(sizeStr.substr(0, xPos));
+                                            if (size > largestSize) {
+                                                largestSize = size;
+                                                largestIconPath = path;
+                                            }
+                                        } catch (...) {
+                                            // Skip invalid size format
+                                            continue;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
```

Comment:
This block of code is really hard to read and understand and does not actually validate the path structure. But given the amount of nested functionality, this also should be a separate function.